auto_ria/auto_ria/spiders/ria_spider.py

```diff
@@ -9,25 +10,41 @@ class RiaSpider(scrapy.Spider):
     start_urls = []
     page_counter = 1
 
-    def __init__(self, url=None, *args, **kwargs):
+    def __init__(self, url=None, f=None, *args, **kwargs):
         super(RiaSpider, self).__init__(*args, **kwargs)
         self.search_url = url
+        self.input_file = f
+
+    def handle_params(self, url, filename):
+        if filename is not None:
```

Comment:
This `if - else` completely changes the class behavior, so consider splitting this into two or three classes - one could work with CSV files, another with URLs and the third one would implement the common behavior.

It can be (better) a structure based on composition (all classes are independent, CSV / URL handlers use the third class) or based on inheritance where common behavior is in the base class and sub-classes implement CSV / URL handling.

Reply:
To be honest I don't know how to re-structure the code this way... For me it seems that common behavior = url behavior, and the only thing that comes to my mind is to write something like `CsvHandler()` class, but it would only contain one method `parse_csv()`.
Please have a look at the attached (pretty rough) code scheme and explain a bit more about this restructure :roll_eyes: 
![scrapy_auto_ria](https://user-images.githubusercontent.com/24430812/29970719-4e7baa08-8f2e-11e7-85cf-2120e2e732ea.png)

Reply:
Wow! That's quite impressive!
Actually, current code is fine, the structure I proposed is just what I would consider better OOP approach (and it would become useful in a more complex case).

What I mean is something like this:

```
class RiaUrls:

    def __init__(self, url):
        self._url = url
        self._start_urls = []

    def scan(self):
        self.logger.debug('Analysing url')
        search_regex = re.compile(r'https://auto.ria.com/search.*')
        single_regex = re.compile(r'https://auto.ria.com/auto.*')
        if re.search(search_regex, url) is not None:
            # means that we have an url to the search
            self.logger.debug('search url match')
            self._find_urls(url)
        elif re.search(single_regex, url) is not None:
            # url to a single item
            self.logger.debug('single url match')
            self._append_url(url)
        else:
            self.logger.debug('no url found')
        return self._start_urls

    def _find_urls(self, url):
        """Find all ads urls inside url till the end."""
        headers = {
            'accept-encoding': 'gzip, deflate, br',
            'accept-language':
            'en-US,en;q=0.8,ru;q=0.6',
            'upgrade-insecure-requests': '1',
            'user-agent':
                'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36',
            'accept':
                'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
            'cookie': 'cookie: _ym_uid=1484309026450282985; showNewFeatures=7; showNewFeaturesMainPage=4; left_filter_test=1;'

        }
        resp = requests.get(url=url, headers=headers)
        soup = BeautifulSoup(resp.text, 'html.parser')
        ads = soup.find_all('a', {'class': 'address'}, href=True)
        unique_urls = []
        for ad in ads:
            if ad['href'] not in unique_urls:
                unique_urls.append(ad['href'])
        if len(unique_urls) > 0:
            self._start_urls.extend(unique_urls)
            next_page_url = '{}&page={}'.format(self.search_url, self.page_counter)
            self.page_counter += 1
            self.find_urls(next_page_url)

    def _append_url(self, url):
        self._start_urls.append(url)


class RiaSpiderUrlHandler:

    def __init__(self, url):
        self._url = url

    def run(self):
        urls = RiaUrls(self._url)
        return urls.scan()


class RiaSpiderFileHandler:

    def __init__(self, filename):
        self._filename = filename

    def run(self):
        urls = []
        with open(self._filename, 'r') as csvfile:
            reader = csv.reader(csvfile)
            for row in reader:
                row_urls = RiaUrls(row[0])
                urls = urls + urls.scan()
        return urls


class RiaSpider(scrapy.Spider):
    name = "auto_ria"
    page_counter = 1

    def __init__(self, url=None, f=None, *args, **kwargs):
        super(RiaSpider, self).__init__(*args, **kwargs)
        if url:
            self.handler = RiaSpiderUrlHandler(url)
        if f:
            self.handler = RiaSpiderFileHandler(f)
        if not self.handler:
            raise Exception('Either `url` or `f` parameter is required')

    def start_requests(self):
        """Use search url and get the list of ads urls."""
        urls = self.handler.run()
        for url in self.start_urls:
            yield self.make_requests_from_url(url)

    def parse(self, response):
        params = [
            ('Link', response.url),
            ('Brand',
                response.css('span[itemprop="brand"]::text').extract_first()),
            ('Model',
                response.css('span[itemprop="name"]::text').extract_first()),
            ('Name',
                response.css('dt.user-name strong::text').extract_first()),
            ('Phone',
                response.css('span.mhide::attr(data-phone-number)').extract_first()),
            ('Price', response.css('span.price::text').extract_first()),
            ('City',
                response.css('span[title="Город продавца"]::text').extract_first()),
            ('Mileage', response.css('div.run strong::text').extract_first()),
            ('Year', response.css('span.year::text').extract_first())
        ]
        yield OrderedDict(params)

```

Reply:
I didn't actually run it, but I think the idea should be clear. And `logger` can be global for module I think.